{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# tNGS Starlims Import\n",
    "\n",
    "## Mutation Surveyor import formats\n",
    "- SNV = c.[x]+[y/=],p.A123A/B\n",
    "- Structural variants = c.123_456_hetdup/hetdel/hetinsACTG or c.123_456_del/dup/ins\n",
    "\n",
    "## Extra Columns\n",
    "- Accession number\n",
    "- Single gene\n",
    "- NucleotideNom\n",
    "- AminoNom\n",
    "- GenomicNom\n",
    "\n",
    "## Pattern logic\n",
    "- If SNV or structural variant\n",
    "- Genotype\n",
    "\n",
    "## Column for mutation details import\n",
    "- GENE p./N c./N"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "df_all = pd.read_csv('data/tNGS_import.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get ride of unwanted columns and create a new df\n",
    "df = df_all.iloc[:,:23].copy()\n",
    "df.rename(columns={'WT nucleotides':'Ref', 'Variant nucleotides':'Alt', 'Variant type':'varType', 'Inserted nucleotides':'insBases'}, inplace=True)\n",
    "sample_ids = list(x for x in df['Folder number'] if x.startswith(\"EX\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Folder number</th>\n",
       "      <th>Report variant?</th>\n",
       "      <th>Reason for selection</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Genomic nomenclature</th>\n",
       "      <th>cDNA nomenclature</th>\n",
       "      <th>Protein nomenclature</th>\n",
       "      <th>Coding effect</th>\n",
       "      <th>varType</th>\n",
       "      <th>...</th>\n",
       "      <th>insBases</th>\n",
       "      <th>Deleted nucleotides</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>AccessionNo</th>\n",
       "      <th>cDNANo</th>\n",
       "      <th>AminoNo</th>\n",
       "      <th>GenomicNo</th>\n",
       "      <th>MutDetails</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>48</td>\n",
       "      <td>snvcontrol05</td>\n",
       "      <td>Variant detected</td>\n",
       "      <td>HGMD sub-category</td>\n",
       "      <td>SRD5A2</td>\n",
       "      <td>0/1</td>\n",
       "      <td>NG_008365.1:g.5367T&gt;C</td>\n",
       "      <td>NM_000348.3:c.281+15T&gt;C</td>\n",
       "      <td>p.?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>substitution</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>NM_000348.3</td>\n",
       "      <td>281+15</td>\n",
       "      <td></td>\n",
       "      <td>5367</td>\n",
       "      <td>SRD5A2 int1 p.?/N c.281+15T&gt;C/N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>49</td>\n",
       "      <td>snvcontrol05</td>\n",
       "      <td>Variant detected</td>\n",
       "      <td>Distance to splice site</td>\n",
       "      <td>CYP11B1</td>\n",
       "      <td>0/1</td>\n",
       "      <td>Chr8(GRCh37):g.143957129G&gt;T</td>\n",
       "      <td>NM_000497.3:c.1120C&gt;A</td>\n",
       "      <td>p.Arg374=</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>substitution</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>NM_000497.3</td>\n",
       "      <td>1120</td>\n",
       "      <td>Arg374=</td>\n",
       "      <td>143957129</td>\n",
       "      <td>CYP11B1 ex6 p.?A?r?g?3?7?4?A?r?g?/N c.1120C&gt;A/N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>50</td>\n",
       "      <td>snvcontrol06</td>\n",
       "      <td>Variant detected</td>\n",
       "      <td>HGMD sub-category</td>\n",
       "      <td>ZMPSTE24</td>\n",
       "      <td>0/1</td>\n",
       "      <td>Chr1(GRCh37):g.40756551dup</td>\n",
       "      <td>NM_005857.4:c.1085dup</td>\n",
       "      <td>p.Leu362Phefs*19</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>duplication</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NM_005857.4</td>\n",
       "      <td>1085</td>\n",
       "      <td>Leu362Phefs*</td>\n",
       "      <td>40756551</td>\n",
       "      <td>ZMPSTE24 ex9 p.?L?e?u?3?6?2?f?s?/N c.1085dup/N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>51</td>\n",
       "      <td>EX1903669</td>\n",
       "      <td>Confirmation required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PHEX_ex_2,PHEX_ex_3</td>\n",
       "      <td>0.661</td>\n",
       "      <td>chrX:22056421-22065408</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>deletion</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X</td>\n",
       "      <td>nan</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>22056421-22065408</td>\n",
       "      <td>PHEX ex2-3 p.? g.22056421-22065408del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>52</td>\n",
       "      <td>EX1903886</td>\n",
       "      <td>Confirmation required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDC73_ex_11,CDC73_ex_12,CDC73_ex_13</td>\n",
       "      <td>0.67</td>\n",
       "      <td>chr1:193172754-193181723</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>deletion</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>nan</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>193172754-193181723</td>\n",
       "      <td>CDC73 ex11-13 p.? g.193172754-193181723del</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Folder number        Report variant?     Reason for selection  \\\n",
       "48  snvcontrol05       Variant detected        HGMD sub-category   \n",
       "49  snvcontrol05       Variant detected  Distance to splice site   \n",
       "50  snvcontrol06       Variant detected        HGMD sub-category   \n",
       "51     EX1903669  Confirmation required                      NaN   \n",
       "52     EX1903886  Confirmation required                      NaN   \n",
       "\n",
       "                                   Gene Genotype         Genomic nomenclature  \\\n",
       "48                               SRD5A2      0/1        NG_008365.1:g.5367T>C   \n",
       "49                              CYP11B1      0/1  Chr8(GRCh37):g.143957129G>T   \n",
       "50                             ZMPSTE24      0/1   Chr1(GRCh37):g.40756551dup   \n",
       "51                  PHEX_ex_2,PHEX_ex_3    0.661       chrX:22056421-22065408   \n",
       "52  CDC73_ex_11,CDC73_ex_12,CDC73_ex_13     0.67     chr1:193172754-193181723   \n",
       "\n",
       "          cDNA nomenclature Protein nomenclature Coding effect       varType  \\\n",
       "48  NM_000348.3:c.281+15T>C                  p.?           NaN  substitution   \n",
       "49    NM_000497.3:c.1120C>A            p.Arg374=    synonymous  substitution   \n",
       "50    NM_005857.4:c.1085dup     p.Leu362Phefs*19    frameshift   duplication   \n",
       "51                      NaN                  NaN           NaN      deletion   \n",
       "52                      NaN                  NaN           NaN      deletion   \n",
       "\n",
       "    ... insBases  Deleted nucleotides  Ref  Alt Chromosome  AccessionNo  \\\n",
       "48  ...      NaN                  NaN    T    C          2  NM_000348.3   \n",
       "49  ...      NaN                  NaN    C    A          8  NM_000497.3   \n",
       "50  ...        T                  NaN  NaN  NaN          1  NM_005857.4   \n",
       "51  ...      NaN                  NaN  NaN  NaN          X          nan   \n",
       "52  ...      NaN                  NaN  NaN  NaN          1          nan   \n",
       "\n",
       "    cDNANo       AminoNo            GenomicNo  \\\n",
       "48  281+15                               5367   \n",
       "49    1120       Arg374=            143957129   \n",
       "50    1085  Leu362Phefs*             40756551   \n",
       "51                          22056421-22065408   \n",
       "52                        193172754-193181723   \n",
       "\n",
       "                                         MutDetails  \n",
       "48                  SRD5A2 int1 p.?/N c.281+15T>C/N  \n",
       "49  CYP11B1 ex6 p.?A?r?g?3?7?4?A?r?g?/N c.1120C>A/N  \n",
       "50   ZMPSTE24 ex9 p.?L?e?u?3?6?2?f?s?/N c.1085dup/N  \n",
       "51            PHEX ex2-3 p.? g.22056421-22065408del  \n",
       "52       CDC73 ex11-13 p.? g.193172754-193181723del  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Accession number\n",
    "df['AccessionNo'] = df['cDNA nomenclature'].apply(lambda x: re.split(\"\\:\", str(x), 1)[0])\n",
    "# cDNA\n",
    "df['cDNANo'] = df['cDNA nomenclature'].apply(lambda x: str(re.findall(\"[^c\\.][0-9]+[+-_]*[0-9]+\", str(x))[-1:]).strip(\"[]''\"))\n",
    "# AminoAccid\n",
    "df['AminoNo'] = df['Protein nomenclature'].apply(lambda x: str(re.findall(\"[a-zA-Z]{3}[0-9]+[\\_]*[a-zA-Z]*[0-9]*[\\=*]*\", str(x))).strip(\"[]''\"))\n",
    "# Genomic\n",
    "df['GenomicNo'] = df['Genomic nomenclature'].apply(lambda x: str(re.findall(\"[0-9]+[\\-]*[0-9]*\", str(x))[-1:]).strip(\"[]''\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Functions\n",
    "def gene_name(gene):\n",
    "    return re.split(\"\\_\", str(gene), 1)[0]\n",
    "\n",
    "def single_amino_code(aa3):\n",
    "    aa_dict = {'Ala': 'A', 'Arg': 'R', 'Asn': 'N', 'Asp': 'D', 'Cys':'C', 'Glu':'E','Gln':'Q', \\\n",
    "               'Gly':'G', 'His':'H', 'Ile':'I','Leu':'L','Lys':'K','Met':'M','Phe':'F','Pro':'P','Ser':'S', \\\n",
    "               'Thr':'T','Trp':'W','Tyr':'Y','Val':'V'}\n",
    "    return aa_dict[aa3]\n",
    "    \n",
    "\n",
    "\n",
    "def zygosity(genotype):\n",
    "    try:\n",
    "        genotype = float(genotype)\n",
    "        if (genotype >= 0.4 and genotype <= 0.6) or (genotype >=1.4 and genotype <= 1.6):\n",
    "            return \"0/1\"\n",
    "        elif genotype <=0.1:\n",
    "            return \"1/1\"\n",
    "        else:\n",
    "            return genotype\n",
    "    except:\n",
    "        return genotype\n",
    "    \n",
    "    \n",
    "def get_exons(gene):\n",
    "    try:\n",
    "        gene = gene.split(\",\")\n",
    "        exons = [re.findall(\"_[0-9]+\", x)[0].replace(\"_\",\"\") for x in gene]\n",
    "    except:\n",
    "        return gene\n",
    "\n",
    "    if len(exons) == 1:\n",
    "        return int(exons[0])\n",
    "    else:\n",
    "        return exons[0] + \"-\" + exons[-1]\n",
    "    \n",
    "    \n",
    "def variant_type(varType, insBases):\n",
    "    if varType == \"duplication\":\n",
    "        return \"dup\"\n",
    "    elif varType == \"deletion\":\n",
    "        return \"del\"\n",
    "    elif varType == \"insertion\":\n",
    "        return \"ins\" + insBases\n",
    "    elif varType == \"delins\":\n",
    "        return \"delins\" + insBases\n",
    "    else:\n",
    "        return \"\"\n",
    "\n",
    "\n",
    "def mutation_details(chrom, gene, exon, intron, amino, cdna, ref, alt, genomic, genotype, varType, insBases):\n",
    "    if pd.isnull(intron):\n",
    "        coding = \"ex\"\n",
    "    else:\n",
    "        coding = \"int\"\n",
    "    \n",
    "    if pd.isnull(exon):\n",
    "        ex_int = get_exons(gene)\n",
    "    else:\n",
    "        try:\n",
    "            ex_int = int(exon)\n",
    "        except:\n",
    "            ex_int = np.nan\n",
    "    gene = gene_name(gene)\n",
    "    genotype = zygosity(genotype)\n",
    "    \n",
    "    varType = variant_type(varType, insBases)\n",
    "    \n",
    "    if pd.isnull(ref) and pd.isnull(exon):\n",
    "        nucleotide = f\"g.{genomic}{varType}\"\n",
    "    elif pd.isnull(ref):\n",
    "        nucleotide = f\"c.{cdna}{varType}\"\n",
    "    else:\n",
    "        nucleotide = f\"c.{cdna}{ref}>{alt}{varType}\"\n",
    "        \n",
    "    try:\n",
    "        amino = amino.replace(\"=\", amino[:3])\n",
    "        amino = re.sub(\"[a-zA-Z]{3}fs\\*\", \"fs\", amino)\n",
    "        #amino = amino.replace(\"\",\"?\")\n",
    "        amino = amino.replace(\"*\", \"Ter\")\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "    if chrom == \"X\" or isinstance(genotype, float):\n",
    "        return f\"{gene} {coding}{ex_int} p.{amino} {nucleotide}\"\n",
    "    elif genotype == \"0/1\":\n",
    "        return f\"{gene} {coding}{ex_int} p.{amino}/N {nucleotide}/N\"\n",
    "    elif genotype == \"1/1\":\n",
    "        return f\"{gene} {coding}{ex_int} p.{amino}/{amino} {nucleotide}/{nucleotide[2:]}\"\n",
    "    \n",
    "    \n",
    "def mut_surveyor(genotype, amino, nucleotide, varType):\n",
    "    nucleotide = nucleotide[2:]  \n",
    "    try:\n",
    "        amino = single_amino_code(re.findall(\"[a-zA-Z]{3}\", amino)[0]) + re.findall(\"[0-9]+\", amino)[0] \\\n",
    "                + single_amino_code(re.findall(\"[a-zA-Z]{3}\", amino)[1])\n",
    "    except IndexError:\n",
    "        amino = single_amino_code(re.findall(\"[a-zA-Z]{3}\", amino)[0]) + re.findall(\"[0-9]+\", amino)[0] + re.findall(\"fs|\\*\", amino)[0]\n",
    "\n",
    "\n",
    "    if varType == \"\":\n",
    "        if genotype == \"0/1\":\n",
    "            return f\"c.[{nucleotide}]+[=],p.{amino}\"\n",
    "        elif genotype == \"1/1\":\n",
    "            return f\"c.[{nucleotide}]+[{nucleotide}],p.{amino}\"\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['MutDetails'] = df.apply(lambda x: mutation_details(x.Chromosome, x.Gene, x.Exon, x.Intron, x.AminoNo, x.cDNANo, x.Ref, x.Alt, x.GenomicNo, x.Genotype, x.varType, x.insBases), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "c.[2116G>A]+[=],p.G706G\n"
     ]
    }
   ],
   "source": [
    "print(mut_surveyor(\"0/1\", 'Gly706Gly', 'c.2116G>A', \"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Folder number                                                  m034\n",
       "Report variant?                                    Variant detected\n",
       "Reason for selection                              HGMD sub-category\n",
       "Gene                                                          ABCC8\n",
       "Genotype                                                        0/1\n",
       "Genomic nomenclature                    Chr11(GRCh37):g.17464303T>C\n",
       "cDNA nomenclature                          NM_001287174.1:c.1594A>G\n",
       "Protein nomenclature                                    p.Ser532Gly\n",
       "Coding effect                                              missense\n",
       "varType                                                substitution\n",
       "Variant location                                               exon\n",
       "Exon                                                            NaN\n",
       "Intron                                                          NaN\n",
       "Quality (vcf)                                               11595.8\n",
       "Filter (vcf)                                                   PASS\n",
       "Allele depth                                                413,360\n",
       "Total depth                                                     773\n",
       "Genotype quality                                                 99\n",
       "insBases                                                        NaN\n",
       "Deleted nucleotides                                             NaN\n",
       "Ref                                                               A\n",
       "Alt                                                               G\n",
       "Chromosome                                                       11\n",
       "AccessionNo                                          NM_001287174.1\n",
       "cDNANo                                                         1594\n",
       "AminoNo                                                   Ser532Gly\n",
       "GenomicNo                                                  17464303\n",
       "MutDetails              ABCC8 ex['ABCC8'] p.Ser532Gly/N c.1594A>G/N\n",
       "Name: 38, dtype: object"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[38]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SNV or Structural variant\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'min'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = \"amino\"\n",
    "a[1:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
